streams/cluster/src/main/java/com/hortonworks/streamline/streams/cluster/service/EnvironmentService.java

```diff
@@ -434,6 +439,85 @@ public NamespaceServiceClusterMapping addOrUpdateServiceClusterMapping(
         return newMapping;
     }
 
+    public void injectProtocolAndPortToComponent(Map<String, String> configurations, Component component) {
+        try {
+            ComponentPropertyPattern confMap = ComponentPropertyPattern
+                    .valueOf(component.getName());
+            String value = configurations.get(confMap.getConnectionConfName());
+            if (value != null) {
+                Matcher matcher = confMap.getParsePattern().matcher(value);
+
+                if (matcher.matches()) {
+                    String protocol = matcher.group(1);
+                    String portStr = matcher.group(2);
+
+                    if (!protocol.isEmpty()) {
+                        component.setProtocol(protocol);
+                    }
+                    if (!portStr.isEmpty()) {
+                        try {
+                            component.setPort(Integer.parseInt(portStr));
+                        } catch (NumberFormatException e) {
+                            LOG.warn(
+                                    "Protocol/Port information [{}] for component {} doesn't seem to known format [{}]."
+                                            + "skip assigning...", value, component.getName(), confMap.getParsePattern());
+
+                            // reset protocol information
+                            component.setProtocol(null);
+                        }
+                    }
+                } else {
+                    LOG.warn("Protocol/Port information [{}] for component {} doesn't seem to known format [{}]. "
+                            + "skip assigning...", value, component.getName(), confMap.getParsePattern());
+                }
+            } else {
+                LOG.warn("Protocol/Port related configuration ({}) is not set", confMap.getConnectionConfName());
+            }
+        } catch (IllegalArgumentException e) {
+            // don't know port related configuration
+        }
+    }
+
+    public Collection<ServiceBundle> listServiceBundles(List<QueryParam> params) {
+        return dao.find(SERVICE_BUNDLE_NAMESPACE, params);
+    }
+
+    public ServiceBundle getServiceBundleByName(String serviceName) {
+        Collection<ServiceBundle> serviceBundles = listServiceBundles(Lists.newArrayList(new QueryParam("name", serviceName)));
+        if (serviceBundles.size() > 1) {
+            LOG.warn("Multiple service bundles have same name: {} returning first match.", serviceName);
+            return serviceBundles.iterator().next();
+        } else if (serviceBundles.size() == 1) {
+            return serviceBundles.iterator().next();
+        }
+        return null;
```

Comment:
The problem is that most DAO methods currently return null and in my opinion they are all incorrect. They should all throw an unchecked exception. For example if a method like getComponentByName, getServiceByName, getServiceConfigurationByName returns null, it means something unexpected occurred. You were looking for something that wasn't there. The client code should be alerted for that, not simply return a null value. A null value is simply a default value. Why is null different from -1, or a string "Not Found" ? For example, the Java Persistence API, when looking up an entity in an underlying data storage (database) and that entity is not found, it throws [EntityNotFoundException](https://docs.oracle.com/javaee/7/api/javax/persistence/EntityNotFoundException.html)